### PR TITLE
Add Python 3 support for Boost package

### PR DESCRIPTION
**Pull request by frol**
- [x] Added support of Python 3 (It has to be built in Python 2 and Python 3 environments now.)
- [x] Fixed some configuration options for OS X and Linux.
- [x] Upgraded to 1.58.0
- [x] Tested on CentOS 6.6, Ubuntu 14.04, OS X
- [x] Add `icu` support for OS X

### CODE REVIEW

File: boost/meta.yaml [hunk 1]
```diff
@@ -17,6 +17,8 @@ requirements:
     - libpython         [win]
```
Comment (ccordoba12) [inline boost/meta.yaml]: Why `python` is not a build dependency?

File: boost/build.sh [hunk 2]
```diff
@@ -8,36 +8,54 @@
...
+        | tee bootstrap.log 2>&1
+
+    ./b2 -q -d0 \
+        variant=release \
+        address-model="${ARCH}" \
+        architecture=x86 \
+        debug-symbols=off \
+        threading=multi \
+        runtime-link=shared \
+        link=shared,static \
```
Comment (ccordoba12) [inline boost/build.sh]: Why do we need to create static libs too?

File: boost/meta.yaml (hunk 1 above)
Comment (frol) [inline boost/meta.yaml]: What do you mean? The first line in build dependencies is `python`. Nevertheless, I haven't touched Windows-specific configs at all. I tested everything on CentOS 6.6, Ubuntu 14.04, OS X, though.

File: boost/build.sh [hunk 3]
```diff
@@ -8,36 +8,54 @@
...
+        toolset=gcc \
+        python="${PY_VER}" \
+        cflags="-O3 -I${PREFIX}/include" \
```
Comment (ccordoba12) [inline boost/build.sh]: Why not use the `include` switch instead of adding build includes to `cflags / cxxflags`?

It seems a bit ugly to me :-)

Comment (ccordoba12) [inline boost/build.sh]: Also, why do you want to optimize this with `03`. Such optimizations can cause problems.

File: boost/build.sh [hunk 4]
```diff
@@ -8,36 +8,54 @@
...
+        cxxflags="-O3 -I${PREFIX}/include" \
+        linkflags="-L${PREFIX}/lib" \
+        --layout=system \
```
Comment (ccordoba12) [inline boost/build.sh]: What does this `layout` switch do?

File: boost/build.sh (hunk 2 above)
Comment (frol) [inline boost/build.sh]: Why shouldn't we? Anyway, I can change it back to shared-only. I copied these options from Arch Linux packaging.

File: boost/meta.yaml (hunk 1 above)
Comment (ccordoba12) [inline boost/meta.yaml]: Ah, sorry. I haven't noticed it. I think it wasn't there before, so I don't know why is not reflected on the diff :-)

File: boost/build.sh (hunk 2 above)
Comment (ccordoba12) [inline boost/build.sh]: I say it because it makes the package bigger for no good reason.

File: boost/build.sh [hunk 5]
```diff
@@ -8,36 +8,54 @@
...
-    variant=release address-model=${ARCH} architecture=x86 \
-    threading=multi link=shared toolset=gcc include=${INCLUDE_PATH} \
-    -j${CPU_COUNT} \
-    install | tee b2.log 2>&1
+if [ "$(uname)" == "Linux" ]; then
+    ./bootstrap.sh \
+        --prefix="${PREFIX}" \
+        --with-python="${PYTHON}" \
+        --with-python-root="${PREFIX} : ${PREFIX}/include/python${PY_VER}m ${PREFIX}/include/python${PY_VER}" \
+        --with-icu="${PREFIX}" \
```
Comment (ccordoba12) [inline boost/build.sh]: Have you checked that `icu` is detected correctly?

File: boost/build.sh (hunk 3 above)
Comment (frol) [inline boost/build.sh]: `--include` switch doesn't work like you think (I expected it would work like that also, BTW). It specifies where Boost will install (!) the libraries, not where it will search on build step. The same thing for `--libsdir`.

Comment (ccordoba12) [inline boost/build.sh]: Well, i used it before to make Boost detect the icu headers.

Comment (frol) [inline boost/build.sh]: `O3` is taken from Arch Linux package, which seems to work for a number of people. Let's optimize as much as possible unless we get a bug report here, this is my way of thinking.

Comment (ccordoba12) [inline boost/build.sh]: Problem is optimizations bugs are hard to detect and/or report. Well, `03` shouldn't cause much harm, but for things to work reliably everything should be compiled with `03` (at least that's how Gentoo works :-).

Since we're not doing that, I'd prefer to **not** add optimizations at all.

File: boost/build.sh [hunk 6]
```diff
@@ -8,36 +8,54 @@
...
-mkdir -vp ${PREFIX}/bin;
+set -x -e
 
-if [ `uname` == Darwin ]; then
+if [ "$(uname)" == "Darwin" ]; then
     MACOSX_VERSION_MIN=10.8
     CXXFLAGS="-mmacosx-version-min=${MACOSX_VERSION_MIN}"
     CXXFLAGS="${CXXFLAGS} -std=c++11 -stdlib=libc++"
     LINKFLAGS="-mmacosx-version-min=${MACOSX_VERSION_MIN} "
-    LINKFLAGS="${LINKFLAGS} -stdlib=libc++ -L${LIBRARY_PATH}"
```
Comment (ccordoba12) [inline boost/build.sh]: Why did you remove this?

File: boost/build.sh [hunk 7]
```diff
@@ -8,36 +8,54 @@
...
+    LINKFLAGS="${LINKFLAGS} -stdlib=libc++"
 
     ./bootstrap.sh \
-      --prefix="${PREFIX}/" --libdir="${PREFIX}/lib/" \
```
Comment (ccordoba12) [inline boost/build.sh]: And also removed this?

File: boost/build.sh (hunk 4 above)
Comment (frol) [inline boost/build.sh]: It is actually `system` by default on Linux anyway, but I just stated it explicitly.

> system - Builds an install without the Boost standard names, and does not install includes to a versioned sub-directory. This is intended for system integrators to build for packaging of distributions.

You can read all options from boost sources or here is dated help that I could find online: http://www.boost.org/doc/libs/1_33_1/Jamfile

File: boost/build.sh (hunk 5 above)
Comment (frol) [inline boost/build.sh]: Yes, it is. Once I specified `"$PREFIX"` explicitly, it gets detected correctly.

File: boost/build.sh (hunk 6 above)
Comment (frol) [inline boost/build.sh]: Because conda-build doesn't provide any LIBRARY_PATH env variable anyway.

File: boost/build.sh (hunk 7 above)
Comment (frol) [inline boost/build.sh]: `libdir` specifies where libraries will be installed, but if you set `prefix` it automatically set `libdir` to `${PREFIX}/lib` and `include` to `${PREFIX}/include`

File: boost/build.sh (hunk 3 above)
Comment (frol) [inline boost/build.sh]: Should I remove `O3` completely, or downgrade to `O2` will be fine?

Comment (frol) [inline boost/build.sh]: @ccordoba12 I have no idea how `include` helped you with `icu` location. Here is the help:

```
Usage: ./bootstrap.sh [OPTION]...

Defaults for the options are specified in brackets.

Configuration:
  -h, --help                display this help and exit
  --with-bjam=BJAM          use existing Boost.Jam executable (bjam)
                            [automatically built]
  --with-toolset=TOOLSET    use specific Boost.Build toolset
                            [automatically detected]
  --show-libraries          show the set of libraries that require build
                            and installation steps (i.e., those libraries
                            that can be used with --with-libraries or
                            --without-libraries), then exit
  --with-libraries=list     build only a particular set of libraries,
                            describing using either a comma-separated list of
                            library names or "all"
                            [all]
  --without-libraries=list  build all libraries except the ones listed []
  --with-icu                enable Unicode/ICU support in Regex
                            [automatically detected]
  --without-icu             disable Unicode/ICU support in Regex
  --with-icu=DIR            specify the root of the ICU library installation
                            and enable Unicode/ICU support in Regex
                            [automatically detected]
  --with-python=PYTHON      specify the Python executable [python]
  --with-python-root=DIR    specify the root of the Python installation
                            [automatically detected]
  --with-python-version=X.Y specify the Python version as X.Y
                            [automatically detected]

Installation directories:
  --prefix=PREFIX           install Boost into the given PREFIX
                            [/usr/local]
  --exec-prefix=EPREFIX     install Boost binaries into the given EPREFIX
                            [PREFIX]

More precise control over installation directories:
  --libdir=DIR              install libraries here [EPREFIX/lib]
  --includedir=DIR          install headers here [PREFIX/include]
```

It states clearly, that it is used only for a deployment of resulting files.

File: boost/build.sh (hunk 4 above)
Comment (ccordoba12) [inline boost/build.sh]: Ok, no problem on leaving it :-)

File: boost/build.sh (hunk 3 above)
Comment (ccordoba12) [inline boost/build.sh]: I used `--include` not `--includedir`, as you can see in the diff :-)

Comment (ccordoba12) [inline boost/build.sh]: Also, please remove all optimizations, i.e. not `03` nor `02`

File: boost/build.sh (hunk 7 above)
Comment (ccordoba12) [inline boost/build.sh]: Ok, thanks for the explanation

File: boost/build.sh (hunk 6 above)
Comment (ccordoba12) [inline boost/build.sh]: Right, I always forget that!

File: boost/build.sh (hunk 2 above)
Comment (ccordoba12) [inline boost/build.sh]: So, please remove `static` libs.

File: boost/build.sh (hunk 3 above)
Comment (frol) [inline boost/build.sh]: `include` is indeed available, so I changed it back. I also removed `O3` and `static`.

File: boost/build.sh (hunk 2 above)
Comment (frol) [inline boost/build.sh]: Done

File: boost/build.sh [hunk 8]
```diff
@@ -8,36 +8,58 @@
...
+
+    ./b2 -q \
+        variant=release \
+        address-model=64 \
+        architecture=x86 \
+        debug-symbols=off \
+        threading=multi \
+        link=shared \
+        toolset=clang \
+        include="${INCLUDE_PATH}" \
```
Comment (ccordoba12) [inline boost/build.sh]: Remember that `INCLUDE_PATH` is not set anymore. This has to be `"${PREFIX}"/include`

Comment (ccordoba12) [inline boost/build.sh]: Sorry, I just saw you set it above.

**Comment by ccordoba12** [comment]
@frol, thanks for the improvements. I'll test this PR this afternoon before merging.

File: boost/build.sh (hunk 3 above)
Comment (frol) [inline boost/build.sh]: Question about optimizations bothered me so I created another issue to discuss it: #376

**Comment by frol** [comment]
@ccordoba12 Have you had a chance to test it? Have you run into any issues?

**Comment by ccordoba12** [comment]
Could you please add the same options for `icu` and `python` that you're using for Linux to Mac too?

I think that's the only thing left before merging.

**Comment by frol** [comment]
@ccordoba12 Those changes were unnecessary for OS X, it seems that autodiscovery worked better there than on Linux. Nevertheless, I set those explicitly now. I also bumped Boost version to 1.59 (released on August 13th). I'm testing the new builds now:

```
$ conda install -c salford_systems boost=1.59
```

**Comment by frol** [comment]
@ccordoba12 All my tests passed. Please, note that my focus is on Linux. I have no project on OS X yet.

**Comment by ccordoba12** [comment]
I have. I tested your recipes and they're working fine :-)

**Comment by frol** [comment]
@ccordoba12 How long does it usually take for a package to get to the official anaconda channel?

**Comment by ccordoba12** [comment]
I'm charge in the Boost packages, but I'm very busy right now. I think I'll do the update in three weeks or a month.
